File: frontend/js/views/lease_buyout.js
// Lease buyout / early-termination settlement generator — PV of remaining rent
// + concessions + fee − reletting recovery, via /calc/lease-buyout.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderLeaseBuyout(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.buyout.h1.title">// LEASE BUYOUT SETTLEMENT</span></h1>
        <p class="muted small" data-i18n="view.buyout.hint.intro">
            The financial settlement a tenant pays to be released from a lease early. It computes the present
            value of the remaining base-rent stream, plus unamortized concessions (TI and commissions not yet
            earned back) and a termination fee, less the landlord's expected reletting recovery. Drafting aid,
            not legal/financial advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.buyout.h2.inputs">Buyout inputs</h2>
            <form id="buyout-form" class="inline-form">
                <label><span data-i18n="view.buyout.label.state">State</span>
                    <input type="text" name="state" value="Delaware" required></label>
                <label><span data-i18n="view.buyout.label.landlord">Landlord</span>
                    <input type="text" name="landlord_name" value=""></label>
                <label><span data-i18n="view.buyout.label.tenant">Tenant</span>
                    <input type="text" name="tenant_name" value=""></label>
                <label><span data-i18n="view.buyout.label.property">Premises</span>
                    <input type="text" name="property_label" value="Suite 900"></label>
                <label><span data-i18n="view.buyout.label.rent">Monthly rent ($)</span>
                    <input type="number" step="100" min="0" name="monthly_rent_usd" value="10000" required></label>
                <label><span data-i18n="view.buyout.label.months">Months remaining</span>
                    <input type="number" step="1" min="0" name="remaining_months" value="24" required></label>
                <label><span data-i18n="view.buyout.label.discount">Discount rate (%/yr)</span>
                    <input type="number" step="0.1" min="0" name="annual_discount_pct" value="6"></label>
                <label><span data-i18n="view.buyout.label.concessions">Unamortized concessions ($)</span>
                    <input type="number" step="1000" min="0" name="unamortized_concessions_usd" value="20000"></label>
                <label><span data-i18n="view.buyout.label.fee">Termination fee (months)</span>
                    <input type="number" step="0.5" min="0" name="termination_fee_months" value="3"></label>
                <label><span data-i18n="view.buyout.label.reletting">Reletting recovery ($)</span>
                    <input type="number" step="1000" min="0" name="reletting_recovery_usd" value="40000"></label>
                <label><span data-i18n="view.buyout.label.date">Date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.buyout.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.buyout.ph.statute'))}"></label>
            </form>
        </div>
        <div id="buyout-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#buyout-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            landlord_name: (fd.get('landlord_name') || '').trim(),
            tenant_name: (fd.get('tenant_name') || '').trim(),
            property_label: (fd.get('property_label') || '').trim(),
            monthly_rent_usd: Number(fd.get('monthly_rent_usd')) || 0,
            remaining_months: Number(fd.get('remaining_months')) || 0,
            annual_discount_pct: Number(fd.get('annual_discount_pct')) || 0,
            unamortized_concessions_usd: Number(fd.get('unamortized_concessions_usd')) || 0,
            termination_fee_months: Number(fd.get('termination_fee_months')) || 0,
            reletting_recovery_usd: Number(fd.get('reletting_recovery_usd')) || 0,
            date: fd.get('date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcLeaseBuyout(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.buyout.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#buyout-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.buyout.card.settlement">Buyout settlement</div>
                    <div class="value">${money(doc.settlement_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.buyout.card.pv">PV remaining rent</div>
                    <div class="value">${money(doc.pv_remaining_rent_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.buyout.card.fee">Termination fee</div>
                    <div class="value">${money(doc.termination_fee_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.buyout.card.concessions">Concessions</div>
                    <div class="value">${money(doc.unamortized_concessions_usd)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.buyout.card.reletting">Less reletting</div>
                    <div class="value">${money(doc.reletting_recovery_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="buyout-copy" type="button" data-i18n="view.buyout.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="buyout-download" type="button" data-i18n="view.buyout.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#buyout-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.buyout.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.buyout.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#buyout-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'lease-buyout.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/lease_buyout.rs
//! Lease buyout / early-termination settlement — the financial settlement a
//! tenant pays to be released from a lease early. Distinct from the
//! lease-termination notice (which is the letter): this computes the settlement
//! figure as the present value of the remaining base-rent stream (an annuity
//! discounted at a monthly rate), plus unamortized concessions (tenant
//! improvements and leasing commissions not yet earned back), plus a termination
//! fee, less the landlord's expected reletting recovery. No existing generator
//! assembles this multi-component buyout. Drafting aid, not legal/financial advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct LeaseBuyoutInput {
    pub landlord_name: String,
    pub tenant_name: String,
    #[serde(default)]
    pub property_label: String,
    /// Monthly base rent remaining on the lease.
    pub monthly_rent_usd: f64,
    /// Months remaining on the term.
    pub remaining_months: u32,
    /// Annual discount rate for present-valuing the remaining rent, percent.
    #[serde(default)]
    pub annual_discount_pct: f64,
    /// Unamortized concessions (TI + leasing commissions) the landlord recovers.
    #[serde(default)]
    pub unamortized_concessions_usd: f64,
    /// Termination fee expressed in months of rent.
    #[serde(default)]
    pub termination_fee_months: f64,
    /// Expected recovery from re-letting the space, which offsets the settlement.
    #[serde(default)]
    pub reletting_recovery_usd: f64,
    pub date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct LeaseBuyout {
    pub title: String,
    /// Nominal remaining rent (monthly × months), undiscounted.
    pub nominal_remaining_rent_usd: f64,
    /// Present value of the remaining rent stream.
    pub pv_remaining_rent_usd: f64,
    /// Termination fee (months × monthly rent).
    pub termination_fee_usd: f64,
    pub unamortized_concessions_usd: f64,
    pub reletting_recovery_usd: f64,
    /// PV rent + concessions + fee − reletting recovery, floored at 0.
    pub settlement_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

/// Present-value annuity factor for `n` monthly payments at monthly rate `r`.
/// With r = 0 the factor is just n (no discounting).
fn annuity_factor(r: f64, n: u32) -> f64 {
    if r > 0.0 {
        (1.0 - (1.0 + r).powi(-(n as i32))) / r
    } else {
        n as f64
    }
}

pub fn generate(i: &LeaseBuyoutInput) -> LeaseBuyout {
    let nominal = cents(i.monthly_rent_usd * i.remaining_months as f64);
    let r = i.annual_discount_pct / 100.0 / 12.0;
    let pv_rent = cents(i.monthly_rent_usd * annuity_factor(r, i.remaining_months));
    let fee = cents(i.termination_fee_months * i.monthly_rent_usd);
    let settlement = cents(
        (pv_rent + i.unamortized_concessions_usd + fee - i.reletting_recovery_usd).max(0.0),
    );

    let property = if i.property_label.trim().is_empty() {
        "the Premises".to_string()
    } else {
        i.property_label.trim().to_string()
    };

    let disc_desc = if i.annual_discount_pct > 0.0 {
        format!("present-valued at {:.2}% per year", i.annual_discount_pct)
    } else {
        "undiscounted".to_string()
    };

    let calc_body = format!(
        "Remaining rent of {} ({} months × {}) {} is {}. Adding unamortized concessions of {} and a termination fee of {} ({:.1} months' rent), and subtracting expected reletting recovery of {}, the buyout settlement is {}.",
        money(nominal),
        i.remaining_months,
        money(i.monthly_rent_usd),
        disc_desc,
        money(pv_rent),
        money(i.unamortized_concessions_usd),
        money(fee),
        i.termination_fee_months,
        money(i.reletting_recovery_usd),
        money(settlement)
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("This settlement is governed by the lease and the laws of the State of {}.", i.state)
    } else {
        format!("This settlement is governed by the lease and the laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Landlord: {}\nTenant: {}\nPremises: {}\nDate: {}",
                i.landlord_name, i.tenant_name, property, i.date
            ),
        },
        DocClause {
            heading: "1. Early Termination".into(),
            body: format!(
                "The Tenant requests to terminate the lease early, with {} months remaining at a monthly base rent of {}. In consideration of release, the Tenant shall pay the buyout settlement computed below.",
                i.remaining_months, money(i.monthly_rent_usd)
            ),
        },
        DocClause { heading: "2. Settlement Calculation".into(), body: calc_body },
        DocClause {
            heading: "3. Payment & Release".into(),
            body: format!(
                "Upon payment of {} and surrender of the Premises in the required condition, the Landlord releases the Tenant from further rent and obligations accruing after the termination date.",
                money(settlement)
            ),
        },
        DocClause {
            heading: "4. Mitigation".into(),
            body: "The reletting recovery reflects the Landlord's expected mitigation. If the Landlord relets sooner or at higher rent than assumed, the parties have nonetheless agreed to the fixed settlement above in full satisfaction.".into(),
        },
        DocClause { heading: "5. Governing Law".into(), body: governing },
        DocClause {
            heading: "Signatures".into(),
            body: format!(
                "Landlord: ____________________  Date: __________\n{}\n\nTenant: ____________________  Date: __________\n{}",
                i.landlord_name, i.tenant_name
            ),
        },
    ];

    LeaseBuyout {
        title: "Lease Buyout / Early Termination Settlement".into(),
        nominal_remaining_rent_usd: nominal,
        pv_remaining_rent_usd: pv_rent,
        termination_fee_usd: fee,
        unamortized_concessions_usd: cents(i.unamortized_concessions_usd),
        reletting_recovery_usd: cents(i.reletting_recovery_usd),
        settlement_usd: settlement,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.02
    }

    fn base() -> LeaseBuyoutInput {
        LeaseBuyoutInput {
            landlord_name: "Tower Owners LP".into(),
            tenant_name: "Departing Tenant LLC".into(),
            property_label: "Suite 900".into(),
            monthly_rent_usd: 10_000.0,
            remaining_months: 24,
            annual_discount_pct: 6.0,
            unamortized_concessions_usd: 20_000.0,
            termination_fee_months: 3.0,
            reletting_recovery_usd: 40_000.0,
            date: "2026-07-01".into(),
            state: "Delaware".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn pv_and_settlement() {
        let d = generate(&base());
        assert!(close(d.nominal_remaining_rent_usd, 240_000.0));
        // PV of 24 × 10,000 at 0.5%/mo ≈ 225,628.66.
        assert!(close(d.pv_remaining_rent_usd, 225_628.66));
        assert!(close(d.termination_fee_usd, 30_000.0));
        // 225,628.66 + 20,000 + 30,000 − 40,000 = 235,628.66.
        assert!(close(d.settlement_usd, 235_628.66));
    }

    #[test]
    fn zero_discount_uses_nominal() {
        let d = generate(&LeaseBuyoutInput { annual_discount_pct: 0.0, ..base() });
        assert!(close(d.pv_remaining_rent_usd, 240_000.0));
        // 240,000 + 20,000 + 30,000 − 40,000 = 250,000.
        assert!(close(d.settlement_usd, 250_000.0));
    }

    #[test]
    fn pv_less_than_nominal_when_discounted() {
        let d = generate(&base());
        assert!(d.pv_remaining_rent_usd < d.nominal_remaining_rent_usd);
    }

    #[test]
    fn settlement_floored_at_zero() {
        // Huge reletting recovery would drive the settlement negative; floor at 0.
        let d = generate(&LeaseBuyoutInput { reletting_recovery_usd: 1_000_000.0, ..base() });
        assert!(close(d.settlement_usd, 0.0));
    }

    #[test]
    fn no_fee_no_concessions() {
        let d = generate(&LeaseBuyoutInput {
            termination_fee_months: 0.0,
            unamortized_concessions_usd: 0.0,
            reletting_recovery_usd: 0.0,
            ..base()
        });
        assert!(close(d.termination_fee_usd, 0.0));
        assert!(close(d.settlement_usd, d.pv_remaining_rent_usd));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&LeaseBuyoutInput { statute_citation: "lease § 21".into(), ..base() });
        assert_eq!(d.statutory_citation, "lease § 21");
        assert!(d.clauses.iter().any(|c| c.body.contains("lease § 21")));
    }
}
